rescan.go

```diff
@@ -573,25 +538,50 @@ func rescan(chain ChainSource, options ...RescanOption) error {
 		// If the filter is nil, then this either means that we don't
 		// have any peers to fetch this filter from, or the peer(s) that
 		// we're trying to fetch from are in the progress of a re-org.
+		// In any case, we'll attempt to retry fetching the filter until
+		// success or we determine it belongs to a stale block.
 		if blockFilter == nil {
-			// TODO(halseth): this is racy, as blocks can come in
-			// before we refetch.
-			resetBlockReFetchTimer(header, uint32(curStamp.Height))
+			// Only add it to the queue if it hasn't been added yet.
+			if !isRetry {
+				retryBlocks = append(retryBlocks, ntfn)
+			}
+
+			log.Debugf("Retrying filter fetch for height=%v, "+
+				"hash=%v after %v", ntfn.Height(),
+				newHeader.BlockHash(), blockRetryInterval)
+
+			blockRetrySignal = time.After(blockRetryInterval)
+
 			return nil
 		}
 
+		// Otherwise, we've obtained the filter, so we'll proceed to
+		// retrieve any matches and notify them.
 		err = notifyBlockWithFilter(
-			chain, ro, &curHeader, &curStamp, blockFilter,
+			chain, ro, &newHeader, &newStamp, blockFilter,
 		)
 		if err != nil {
 			return err
 		}
 
-		// We've successfully fetched this current block, so we'll reset
-		// the retry timer back to nil.
-		if blockReFetchTimer != nil {
-			blockReFetchTimer.Stop()
-			blockReFetchTimer = nil
+		// With the notifications sent, we can update our in-memory
+		// state.
+		curHeader = newHeader
+		curStamp = newStamp
+
+		// If this was a retry, we'll proceed to the next if any.
+		if isRetry {
+			retryBlocks = retryBlocks[1:]
```

Comment:
How do we handle the following case:
  * We're trying to fetch a new block, however a re-org occurs at that point 
  * As a result, if we try to fetch that blocks, we'll _never_ get a response since it's no longer in the main-chain (`btcd`'s behavior) 

In this case, it would seem that we'll need to _replace_ newer items for older items within the retry queue if we get two at the same height.

Reply:
We could definitely do that, though the assumption is that we'll receive block disconnected notifications from the `blockManager` for those blocks, which we'll then process within `handleBlockDisconnected` and remove them. Given that, do you still think it's worth handling this case explicitly here?

Reply:
> We could definitely do that, though the assumption is that we'll receive block disconnected notifications from the blockManager for those blocks, which we'll then process within handleBlockDisconnected and remove them

After reading the rest of this diff, it seems this case is addressed. However, the scenario I described above can still happen if _right_ as we go to fetch that block, a re-org happens.